qiskit/algorithms/eigensolvers/vqd.py

```diff
@@ -0,0 +1,431 @@
+            This is the objective function to be passed to the optimizer that is used for evaluation.
+
+        Args:
+            step: level of energy being calculated. 0 for ground, 1 for first excited state...
+            operator: The operator whose energy to evaluate.
+            prev_states: List of parameters from previous rounds of optimization.
+
+        Returns:
+            A callable that computes and returns the energy of the hamiltonian
+            of each parameter.
+
+        Raises:
+            RuntimeError: If the circuit is not parameterized (i.e. has 0 free parameters).
+            AlgorithmError: If operator was not provided.
+
+        """
+
+        num_parameters = self.ansatz.num_parameters
+        if num_parameters == 0:
+            raise RuntimeError("The ansatz must be parameterized, but has 0 free parameters.")
+
+        if step > 1 and (len(prev_states) + 1) != step:
+            raise RuntimeError(
+                f"Passed previous states of the wrong size."
+                f"Passed array has length {str(len(prev_states))}"
+            )
+
+        self._check_operator_ansatz(operator)
+
+        prev_circs = []
+        for state in range(step - 1):
+            prev_circs.append(self.ansatz.bind_parameters(prev_states[state]))
+
+        def evaluate_energy(parameters: np.ndarray) -> np.ndarray | float:
+
+            try:
+                estimator_job = self.estimator.run(
+                    circuits=[self.ansatz], observables=[operator], parameter_values=[parameters]
+                )
+                estimator_result = estimator_job.result()
+                means = np.real(estimator_result.values)
+
+            except Exception as exc:
+                raise AlgorithmError("The primitive job to evaluate the energy failed!") from exc
+
+            if step > 1:
+                # Compute overlap cost
+                fidelity_job = self.fidelity.run(
+                    [self.ansatz] * len(prev_circs),
+                    prev_circs,
+                    [parameters] * len(prev_circs),
+                    [prev_states[:-1]],
+                )
+                costs = fidelity_job.result().fidelities
+
+                for (state, cost) in zip(range(step - 1), costs):
+                    means += np.real(self.betas[state] * cost)
+
+            if self.callback is not None:
+                metadata = estimator_result.metadata
+                for params, mean, meta in zip([parameters], means, metadata):
+                    self._eval_count += 1
+                    self.callback(self._eval_count, params, mean, meta, step)
+            else:
+                self._eval_count += len(means)
+
+            return means if len(means) > 1 else means[0]
+
+        return evaluate_energy
+
+    @staticmethod
+    def _build_vqd_result() -> VQDResult:
+
+        result = VQDResult()
+        result.optimal_point = []
```

Comment:
It seems with lists we are going beyond the typehints for VariationalAlgo. And even on the initial point getter/setter though its Sequence float I guess ideally it would be a list of initial points not a single one, though I think here we only support an single value that is used on every step.

Reply:
Removed extension of `VariationalResult`

Reply:
Maybe we need to think about improving initial_point as input gong forwards since I think it ought to be a list of points one for each step, maybe we support a single value, like now for all. But ideally I should be able to feed in the optimal points outcome which is a list of points so each step has its optimizer set accordingly.

Reply:
I also agree, I will note it as a future improvement (again, not resolving the comment to be able to find it later).